### ISSUE TITLE
`BasePredictionWriter` passes `batch_indices` as `List[List[List[int]]]` instead of `List[List[int]]`

### ISSUE THREAD

**Issue by lucadiliello** [Lightning-AI/lightning#17643]
### Bug description

`BasePredictionWriter` passes `batch_indices` as `List[List[List[int]]]` instead of `List[List[int]]` to method `write_on_epoch_end`, while `predictions` is just a `List[Any]`. I expect `batch_indices` to pass a `List[List[int]]` that contains batch indexes for each batch.

Current `batch_indices` shape: `[1, #batches, #batch_size]`
Expected `batch_indices` shape: `[#batches, #batch_size]`

Maybe the additional dimension is used for multiple dataloaders? Then how should I differentiate between different dataloaders if `predictions` is only a `List[Any]` returned by the `predict_step`?

### What version are you seeing the problem on?

v2.0

### Error messages and logs

```
# Error messages and logs here please
```

### Environment

<details>
  <summary>Current environment</summary>

* CUDA:
	- GPU:
		- NVIDIA GeForce RTX 3090
		- NVIDIA GeForce RTX 3090
	- available:         True
	- version:           11.8
* Lightning:
	- bleurt-pytorch:    0.0.1
	- lightning:         2.0.2
	- lightning-cloud:   0.5.34
	- lightning-fabric:  2.0.2
	- lightning-utilities: 0.8.0
	- pytorch-lightning: 2.0.2
	- torch:             2.0.1+cu118
	- torchmetrics:      0.11.4
* Packages:
	- accelerate:        0.18.0
	- aiobotocore:       2.4.2
	- aiohttp:           3.8.3
	- aioitertools:      0.11.0
	- aiosignal:         1.3.1
	- anyio:             3.6.2
	- apache-beam:       2.44.0
	- arrow:             1.2.3
	- asttokens:         2.2.1
	- async-timeout:     4.0.2
	- asyncio:           3.4.3
	- attrs:             22.2.0
	- awscli:            1.27.126
	- awscli-plugin-endpoint: 0.4
	- bcrypt:            4.0.1
	- beautifulsoup4:    4.12.2
	- bitsandbytes:      0.38.1
	- blessed:           1.20.0
	- bleurt-pytorch:    0.0.1
	- blingfire:         0.1.8
	- botocore:          1.29.126
	- certifi:           2022.12.7
	- cffi:              1.15.1
	- cfgv:              3.3.1
	- charset-normalizer: 2.1.1
	- click:             8.1.3
	- cloudpickle:       2.2.1
	- cmake:             3.26.0
	- colorama:          0.4.4
	- commonmark:        0.9.1
	- contexttimer:      0.3.3
	- contourpy:         1.0.6
	- crcmod:            1.7
	- croniter:          1.3.14
	- cryptography:      39.0.0
	- cycler:            0.11.0
	- datasets:          2.12.0
	- dateutils:         0.6.12
	- deepdiff:          6.3.0
	- deepspeed:         0.9.1
	- deepspeed-mii:     0.0.5+9ec2f12
	- dill:              0.3.5.1
	- distlib:           0.3.6
	- docopt:            0.6.2
	- docutils:          0.16
	- exceptiongroup:    1.1.0
	- execnet:           1.9.0
	- executing:         1.2.0
	- fabric:            2.7.1
	- fastapi:           0.88.0
	- fastavro:          1.7.0
	- fasteners:         0.18
	- filelock:          3.9.0
	- fonttools:         4.38.0
	- frozenlist:        1.3.3
	- fsspec:            2023.1.0
	- grpcio:            1.51.3
	- grpcio-tools:      1.51.3
	- h11:               0.14.0
	- hdfs:              2.7.0
	- hjson:             3.1.0
	- httplib2:          0.20.4
	- huggingface-hub:   0.14.1
	- icecream:          2.1.3
	- identify:          2.5.12
	- idna:              3.4
	- iniconfig:         2.0.0
	- inquirer:          3.1.3
	- invoke:            1.7.3
	- itsdangerous:      2.1.2
	- jinja2:            3.1.2
	- jmespath:          1.0.1
	- joblib:            1.2.0
	- kiwisolver:        1.4.4
	- lightning:         2.0.2
	- lightning-cloud:   0.5.34
	- lightning-fabric:  2.0.2
	- lightning-utilities: 0.8.0
	- lit:               15.0.7
	- llvmlite:          0.39.1
	- markupsafe:        2.1.2
	- matplotlib:        3.6.2
	- mkl-fft:           1.3.1
	- mkl-random:        1.2.2
	- mkl-service:       2.4.0
	- mpmath:            1.3.0
	- multidict:         6.0.4
	- multiprocess:      0.70.13
	- networkx:          3.0
	- ninja:             1.11.1
	- nltk:              3.8.1
	- nodeenv:           1.7.0
	- numba:             0.56.4
	- numpy:             1.23.5
	- nvidia-cublas-cu11: 11.10.3.66
	- nvidia-cuda-nvrtc-cu11: 11.7.99
	- nvidia-cuda-runtime-cu11: 11.7.99
	- nvidia-cudnn-cu11: 8.5.0.96
	- objsize:           0.6.1
	- ordered-set:       4.1.0
	- orjson:            3.8.5
	- packaging:         23.0
	- pandas:            1.5.2
	- paramiko:          2.12.0
	- pathlib2:          2.3.7.post1
	- pillow:            9.4.0
	- pip:               22.3.1
	- platformdirs:      2.6.2
	- pluggy:            1.0.0
	- pre-commit:        2.21.0
	- prettytable:       3.6.0
	- proto-plus:        1.22.2
	- protobuf:          3.20.1
	- psutil:            5.9.4
	- py-cpuinfo:        9.0.0
	- pyarrow:           9.0.0
	- pyasn1:            0.4.8
	- pycparser:         2.21
	- pydantic:          1.10.4
	- pydot:             1.4.2
	- pygments:          2.14.0
	- pyjwt:             2.6.0
	- pymongo:           3.13.0
	- pynacl:            1.5.0
	- pyparsing:         3.0.9
	- pytest:            7.2.1
	- pytest-xdist:      3.2.0
	- python-dateutil:   2.8.2
	- python-editor:     1.0.4
	- python-multipart:  0.0.6
	- pytorch-lightning: 2.0.2
	- pytz:              2022.7
	- pyyaml:            5.4.1
	- readchar:          4.0.5
	- regex:             2022.10.31
	- requests:          2.28.1
	- responses:         0.18.0
	- rich:              13.0.1
	- rsa:               4.7.2
	- s3fs:              2023.1.0
	- s3transfer:        0.6.0
	- scipy:             1.10.0
	- sentencepiece:     0.1.97
	- setuptools:        65.6.3
	- six:               1.16.0
	- sniffio:           1.3.0
	- soupsieve:         2.4.1
	- starlette:         0.22.0
	- starsessions:      1.3.0
	- sympy:             1.11.1
	- tensorboardx:      2.5.1
	- tokenizers:        0.13.2
	- tomli:             2.0.1
	- torch:             2.0.1+cu118
	- torchmetrics:      0.11.4
	- tqdm:              4.64.1
	- traitlets:         5.9.0
	- transformers:      4.29.1
	- transformers-framework: 2.0.1
	- triton:            2.0.0
	- typing-extensions: 4.4.0
	- urllib3:           1.26.14
	- uvicorn:           0.22.0
	- virtualenv:        20.17.1
	- wcwidth:           0.2.6
	- websocket-client:  1.5.1
	- websockets:        11.0.2
	- wheel:             0.37.1
	- wrapt:             1.14.1
	- xxhash:            3.2.0
	- yarl:              1.8.2
	- zstandard:         0.19.0
* System:
	- OS:                Linux
	- architecture:
		- 64bit
		- ELF
	- processor:         x86_64
	- python:            3.9.13
	- release:           5.15.0-72-generic
	- version:           #79-Ubuntu SMP Wed Apr 19 08:22:18 UTC 2023

</details>

**Comment by awaelchli** [comment]
This seems to be fixed. The `write_on_epoch_end` gets a list of batches. There is no additional dimension unless multiple dataloaders are used. 

Here is the example I used to verify:


```py
from pprint import pprint

import torch
from torch.utils.data import DataLoader, Dataset

from lightning.pytorch import LightningModule, Trainer
from lightning.pytorch.callbacks import BasePredictionWriter


class RandomDataset(Dataset):
    def __init__(self, size, length):
        self.len = length
        self.data = torch.randn(length, size)

    def __getitem__(self, index):
        return self.data[index]

    def __len__(self):
        return self.len


class Writer(BasePredictionWriter):
    def write_on_batch_end(self, trainer, pl_module, prediction, batch_indices, batch, batch_idx, dataloader_idx):
        print(prediction.shape)
        print(batch_indices)

    def write_on_epoch_end(self, trainer, pl_module, predictions, batch_indices):
        pprint(predictions)


class BoringModel(LightningModule):
    def __init__(self):
        super().__init__()
        self.layer = torch.nn.Linear(32, 2)

    def forward(self, x):
        return self.layer(x)


test_data = DataLoader(RandomDataset(32, 8), batch_size=2)
model = BoringModel()
trainer = Trainer(
    callbacks=[Writer(write_interval="epoch")],
    enable_model_summary=False,
    enable_progress_bar=False,
)
trainer.predict(model, dataloaders=test_data)

```
